Pkg panda integration

abhay-agarwal opened the pull request.

abhay-agarwal commented:
Here's the new changes @elingg @adam-mesos

adam-mesos commented on conf/README.md:
```diff
@@ -1,17 +0,0 @@
-mesos-site.xml
---------------------------
-
-... more documentation to come
```
Why remove these? We should document common/required config params, or at least point to the Hadoop settings for info on hdfs/core-site.xml

abhay-agarwal replied:
elizabeth wanted a single main readme

elingg replied:
That's correct. Could we add the documentation on how the user can use Mesos DNS to our main readme page please? This way it is easier to find and users can immediately read the markdown and figure out how to use Mesos DNS.

adam-mesos commented on conf/mesos-site.xml:
```diff
@@ -99,12 +109,12 @@
   <property>
     <name>mesos.hdfs.journalnode.cpus</name>
-    <value>1</value>
+    <value>0.5</value>
```
Did we do any monitoring/benchmarking to arrive at this number, or is it just smaller so that it's easier to guarantee scheduling on a small/busy cluster?

abhay-agarwal replied:
No work has been done! I'm working on integration tests. For now, I would err on the side of small.

adam-mesos replied:
SGTM. Just want to make sure we're consistent in the xml defaults and the defaults hardcoded into SchedulerConf.java (all the more reason I want to remove or comment out the xml values and leave the defaults to SchedulerConf.java)

elingg replied:
Yes, let's make sure this is consistent with SchedulerConf. @adam-mesos, actually I would like to leave these values here. These are values that users often change. I've had a few users with larger or smaller clusters change them. If it's a value that people are often changing it makes sense to keep them in the config. We might want to add a description.

adam-mesos replied:
If we leave them in, I'd rather we comment them out. That way, the user can still see the description (and default value), and know exactly how to add their own custom value for it. But at the same time we save on xml parsing time. I remember one of the first speedups that MapR made was clearing out the site.xml files, which actually shaved seconds off starting the various Hadoop components.

elingg replied:
I am okay with leaving them in commented out. Sounds good, @adam-mesos

On Wed, Mar 25, 2015 at 9:34 AM, Adam <notifications@github.com> wrote:

> In conf/mesos-site.xml
> <https://github.com/mesosphere/hdfs/pull/103#discussion_r27138692>:
>
> > @@ -99,12 +109,12 @@
> >
> >    <property>
> >      <name>mesos.hdfs.journalnode.cpus</name>
> > -    <value>1</value>
> > +    <value>0.5</value>
>
> If we leave them in, I'd rather we comment them out. That way, the user
> can still see the description (and default value), and know exactly how to
> add their own custom value for it. But at the same time we save on xml
> parsing time. I remember one of the first speedups that MapR made was
> clearing out the site.xml files, which actually shaved seconds off starting
> the various Hadoop components.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mesosphere/hdfs/pull/103/files#r27138692>.
>

adam-mesos commented on example-conf/mesos-dns/mesos-site.xml:
```diff
@@ -0,0 +1,39 @@
+
+<configuration>
+  <property>
+    <name>mesos.hdfs.state.zk</name>
```
What about mesos.hdfs.zkfc.ha.zookeeper.quorum? Or are you just expecting that the hdfs-site.xml is somehow already configured in a mesos-dns setup?

abhay-agarwal replied:
it's based on the config server's default behavior

elingg replied:
mesos.hdfs.zkfc.ha.zookeeper.quorum is set here in mesos-site.xml. @adam-mesos, if it's set in this file, it gets propagated to hdfs-site.xml through this variable, ${haZookeeperQuorum}

adam-mesos commented on example-conf/mesosphere-dcos/hdfs-site.xml:
```diff
@@ -0,0 +1,207 @@
+
+<!-- Put site-specific property overrides in this file. -->
+
+<configuration>
```
This is a lot of config properties. Can any of these be removed so that they just fallback to the default value? I'd rather our config files only call out the differences required from the defaults.

abhay-agarwal replied:
removed from mesos-site. Most of these are necessary.

adam-mesos commented on example-conf/mesosphere-dcos/hdfs-site.xml:
```diff
@@ -0,0 +1,207 @@
+
+  <property>
+    <name>ha.zookeeper.quorum</name>
+    <value>master0.mesos:2181</value>
```
What about multiple zks, should we list master0-2.mesos:2181?

adam-mesos commented on example-conf/mesosphere-dcos/hdfs-site.xml:
```diff
@@ -0,0 +1,207 @@
+
+  <property>
+    <name>dfs.namenode.name.dir</name>
+    <value>file:///tmp/hdfs/data/name</value>
```
Can we come up with a non-/tmp default location that will still work on DCOS? No problem if we can't, since it is configurable if a customer has their own setup.

adam-mesos commented on example-conf/mesos-dns/mesos-site.xml:
```diff
@@ -0,0 +1,39 @@
+
+  <property>
+    <name>mesos.master.uri</name>
+    <value>zk://master.mesos:2181/mesos</value>
```
Can you also provide a <description> tag for these properties you're modifying?

elingg replied:
For the properties that aren't default, a description tag could be useful

adam-mesos commented on example-conf/mesosphere-dcos/hdfs-site.xml:
```diff
@@ -0,0 +1,207 @@
+
+  <property>
+    <name>dfs.datanode.balance.bandwidthPerSec</name>
+    <value>41943040</value>
```
If these are defaults, can we just remove them from the config? If not, where did this come from?

abhay-agarwal replied:
Good question. These defaults have been here since I started working on the project. From git history, these were committed in @brndnmtthws 's original commit. These are all configurations that differ from the defaults.

brndnmtthws replied:
Some of the defaults aren't appropriate for production use (such as this one, which is much too small) hence their inclusion.

adam-mesos commented on example-conf/mesosphere-dcos/mesos-site.xml:
```diff
@@ -0,0 +1,134 @@
+
+<!-- Put site-specific property overrides in this file. -->
+
+<configuration>
```
Again, please remove any default values, and ideally add <description> tags for anything left that they would want to change.

adam-mesos commented on example-conf/mesosphere-dcos/mesos-site.xml:
```diff
@@ -0,0 +1,134 @@
+
+  <property>
+    <name>mesos.hdfs.user</name>
+    <value>root</value>
```
Does this have to be root, or could it be 'hdfs'/'hduser'?

elingg commented:
thanks very much, @abhay-agarwal! I'll take a look ASAP

adam-mesos commented on conf/mesos-site.xml:
```diff
@@ -63,11 +63,21 @@
   </property>
 
   <property>
+    <name>mesos.hdfs.preshared-configuration</name>
```
I'm not a fan of this property name, since you're also using it to trigger symlinks and update the path. How about "preloaded-hadoop" or "preinstalled-hadoop", since you wouldn't need to update symlinks/path or reload hadoop configs if they're already preloaded/preinstalled?

adam-mesos commented:
Great work, @abhay-agarwal! I'd like to see the xml files minimized down to non-defaults and descriptions added, and then I had a few other minor questions/suggestions. Fix/respond to these and I'll give it another look.

elingg commented on conf/hdfs-site.xml:
```diff
@@ -24,72 +24,72 @@
 
   <property>
```
Please don't modify the original conf/hdfs-site.xml file.This will create a bug for users that are not using Mesos DNS. We should leave the original files intact and only add examples for users who want to use Mesos DNS.

elingg commented on conf/mesos-site.xml:
```diff
@@ -18,107 +18,37 @@
 
```
Please don't modify the original conf/mesos-site.xml file.This will create a bug for users that are not using Mesos DNS. We should leave the original files intact and only add examples for users who want to use Mesos DNS.

elingg commented on conf/mesos-site.xml:
```diff
@@ -18,107 +18,37 @@
-    <value>/usr/local/lib/libmesos.so</value>
-  </property>
-
-  <property>
```
We should keep the defaults for memory and cpu here as well. If we need to raise the default values for some of these as it makes sense, that's fine, but we should not remove them. This is an important piece that the user would want to configure.

elingg replied:
It might also be helpful to add a description tag since these are fields they would potentially want to change based on their set up.

elingg commented:
I left my comments above. These are the main things I'd like to see:
1) We don't want to modify the default files (unless there are changes we need to make like updating memory settings). Instead, we want to add new files as examples for the Mesos DNS and Predistributed Config Options. We want to add some comments to make it clear for the user.
2) In our main readme, we want to explain how the user can run HDFS with the Predistributed Config Option and the Mesos DNS Option.

elingg commented on src/main/java/org/apache/mesos/hdfs/config/SchedulerConf.java:
```diff
@@ -34,6 +34,10 @@ public String getMesosDnsDomain() {
     return getConf().get("mesos.hdfs.mesosdns.domain", "mesos");
```
Could you update the default values in schedulerConf to mirror you changes to cpu and memory in mesos-site.xml?

adam-mesos commented on example-conf/mesos-dns/mesos-site.xml:
```diff
@@ -0,0 +1,49 @@
+  <property>
+    <name>mesos.hdfs.zkfc.ha.zookeeper.quorum</name>
+    <value>master0.mesos:2181,master1.mesos:2181,master2.mesos:2181</value>
+    <description>This is where you must individually list all your zookeeper nodes</description>
```
You've explained the format, but how is it used? In this case, this is the list of zk nodes passed to the HDFS tasks via the ConfigServer.

adam-mesos commented on example-conf/mesosphere-dcos/hdfs-site.xml:
```diff
@@ -0,0 +1,207 @@
+
+  <property>
+    <name>dfs.datanode.dns.nameserver</name>
+    <value>default</value>
```
I'm looking at https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-hdfs/hdfs-default.xml and at least interface and nameserver can be excluded, since these are the default values. More can probably also be removed.

elingg replied:
+1 to remove these, if they are actually defaults.

adam-mesos commented on example-conf/mesos-dns/mesos-site.xml:
```diff
@@ -0,0 +1,49 @@
+  <property>
+    <name>mesos.hdfs.state.zk</name>
+    <value>master.mesos:2181</value>
+    <description>Zookeeper is used by Hadoop's HA feature. If you colocate zk and mesos masters, you can use the master location here.</description>
```
This ZK parameter is used by the HDFS framework scheduler itself, for storing state like the frameworkId and task states.
That's separate from the mesos.hdfs.zkfc.ha.zookeeper.quorum variable that the ZKFC uses to contact ZK for HDFS namenode election, which you incorrectly reference here.
By tweaking these parameters, the framework could use a different ZK cluster (e.g. the Mesos ZK) for its persistent state than the ZK cluster that HDFS uses for its own leader election (e.g. the Hadoop ZK).

elingg replied:
True, users may want to use a different zookeeper cluster than the one used by Mesos. We should be clear about that in the description.

adam-mesos commented on example-conf/mesos-dns/mesos-site.xml:
```diff
@@ -0,0 +1,49 @@
+  <property>
+    <name>mesos.hdfs.mesosdns.domain</name>
+    <value>mesos</value>
+    <description>The domain is usually mesos, but if you changed this in Mesos DNS, change it here.</description>
```
s/usually mesos/usually 'mesos'/

adam-mesos commented on example-conf/mesos-dns/mesos-site.xml:
```diff
@@ -0,0 +1,49 @@
+  <property>
+    <name>mesos.master.uri</name>
+    <value>zk://master.mesos:2181/mesos</value>
+    <description>The URI of zookeeper from the above setup is a reference to any of your mesos masters. </description>
```
Not necessarily. You could potentially use a different ZK cluster for scheduler state than what Mesos uses for its leader election.
This flag specifies how to connect to a mesos master or a quorum of masters. See the description for --master in https://mesosphere.github.io/marathon/docs/command-line-flags.html

elingg replied:
True, users may want to use a different zookeeper cluster than the one used by Mesos. We should be clear about that in the description.

adam-mesos commented on README.md:
```diff
@@ -16,6 +16,9 @@ Installing HDFS-Mesos on your Cluster
 3. Customize configuration in `hdfs-mesos-*/etc/hadoop/*-site.xml`
 4. Check that `hostname` on that node resolves to a non-localhost IP; update /etc/hosts if necessary
 
+### Mesos-DNS custom configuration (see example-conf directory for example)
```
No need to say "see example-conf..." since you reference it just below.

adam-mesos commented on README.md:
```diff
@@ -48,3 +51,10 @@ Resource Reservation Instructions (Optional)
+--------------------------
+
+1. In Marathon (or your other long-running process monitor) shutdown the task
+2. Shutdown the mesos framework: `curl -d "frameworkId=YOUR_FRAMEWORK_ID" -X POST http://YOUR_MESOS_URL:5050/master/shutdown`
```
s/Shutdown the mesos framework/Shutdown the hdfs framework in Mesos/

adam-mesos commented on README.md:
```diff
@@ -48,3 +51,10 @@ Resource Reservation Instructions (Optional)
+Shutdown Instructions (Optional)
+--------------------------
+
+1. In Marathon (or your other long-running process monitor) shutdown the task
```
s/shutdown the task/stop the hdfs scheduler application/

adam-mesos commented on README.md:
```diff
@@ -48,3 +51,10 @@ Resource Reservation Instructions (Optional)
+1. In Marathon (or your other long-running process monitor) shutdown the task
+2. Shutdown the mesos framework: `curl -d "frameworkId=YOUR_FRAMEWORK_ID" -X POST http://YOUR_MESOS_URL:5050/master/shutdown`
+3. Access your zookeeper instance: `/PATH/TO/zookeeper/bin/zkCli.sh`
+4. Remove HDFS data from zookeeper: `rmr /hdfs-mesos`
```
s/HDFS data/hdfs-mesos framework state/

adam-mesos commented on conf/mesos-site.xml:
```diff
@@ -63,11 +63,21 @@
 
   <property>
+    <value>false</value>
+    <name>mesos.hdfs.preshared-hadoop-binaries</name>
```
What does "preshared" mean? How about "preinstalled" or "predistributed"? Or reverse the polarity and call it "update-hadoop-install" (default true), or something similar?
Whatever you do, this is a complex enough property to warrant a <description> tag/comment. I'd like to see them for all properties eventually, but we can just add descriptions as we go.

adam-mesos commented on example-conf/mesosphere-dcos/hdfs-site.xml:
```diff
@@ -0,0 +1,207 @@
+  </property>
+
+  <property>
+    <name>dfs.ha.automatic-failover.enabled.hdfs</name>
```
I don't think this is a real property. http://lmgtfy.com/?q=%22dfs.ha.automatic-failover.enabled.hdfs%22
The part just before '.hdfs' is though, but it's covered above.

adam-mesos commented on example-conf/mesosphere-dcos/hdfs-site.xml:
```diff
@@ -0,0 +1,207 @@
+
+  <property>
+    <name>dfs.journalnode.edits.dir</name>
+    <value>/tmp/hdfs/data/jn</value>
```
s/jn/journal/?

adam-mesos replied:
Why is this `/tmp/...` when `dfs.namenode.name.dir` is `file:///tmp/...`?

adam-mesos commented on example-conf/mesosphere-dcos/hdfs-site.xml:
```diff
@@ -0,0 +1,207 @@
+
+  <property>
+    <name>dfs.webhdfs.enabled</name>
+    <value>true</value>
```
Default value, can be removed.

adam-mesos commented on example-conf/mesosphere-dcos/hdfs-site.xml:
```diff
@@ -0,0 +1,207 @@
+
+  <property>
+    <name>dfs.replication</name>
+    <value>3</value>
```
Default, remove.

adam-mesos commented on example-conf/mesosphere-dcos/hdfs-site.xml:
```diff
@@ -0,0 +1,207 @@
+  <property>
+    <name>dfs.namenode.heartbeat.recheck-interval</name>
+    <!-- 60 seconds -->
+    <value>60000</value>
```
This is much shorter than the default 300 seconds. https://issues.apache.org/jira/browse/HDFS-7685
Another one of Brenden's production optimizations?

adam-mesos commented on example-conf/mesosphere-dcos/hdfs-site.xml:
```diff
@@ -0,0 +1,207 @@
+
+  <property>
+    <name>dfs.namenode.handler.count</name>
+    <value>512</value>
```
The number of handler threads you set here could have a big impact on the memory usage of the NN/DNs.

adam-mesos commented on example-conf/mesosphere-dcos/hdfs-site.xml:
```diff
@@ -0,0 +1,207 @@
+
+  <property>
+    <name>dfs.datanode.failed.volumes.tolerated</name>
+    <value>0</value>
```
Default value, can be removed.

adam-mesos commented:
Looking good; I just have some nits about config descriptions and the name "preshared" (where's the "sharing"?). I also found a few more default value configs you can remove. Fix these and it should be just about shippable.
Sorry to nit so much about the descriptions, but I'm trying to consider what's going to help vs. confuse potential users on first glance at our mesos-site.xml and our custom hdfs-site.xml (templates). The subtleties of the different zk parameters in particular are important to note.

abhay-agarwal commented:
Hey @adam-mesos, lets leave the config parameters that Brenden used (except for the default ones which I removed). Then, when we do integration testing with DCOS, we can really tweak these parameters well. As for now, i've updated the configs and changed the names to make things more clear. Let's get this merged soon!

adam-mesos commented:
SGTM. I'll take another look today.
As for Brenden's values, I just wanted to point out that the 512 thread NN could eat lots of memory unnecessarily on a small DCOS cluster (as you previously reported). The rest can be revisited later.

brndnmtthws commented:
Good point Adam. That number was picked for a somewhat large cluster (>500) nodes.

abhay-agarwal commented:
okay will change to 20 for now.

abhay-agarwal commented:
@elingg is this okay to merge?

adam-mesos commented:
@abhay-agarwal I'd like to take another quick look this afternoon, if you don't mind.

abhay-agarwal commented:
all good

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/executor/AbstractNodeExecutor.java:
```diff
@@ -177,9 +180,10 @@ protected void startProcess(ExecutorDriver driver, Task task) {
 
   /**
-   * Reloads the cluster configuration so the executor has the correct configuration info.
+   * Reloads the cluster configuration som the executor has the correct configuration info.
```
s/som/so/

elingg commented:
Thanks for your patience, @abhay-agarwal! @adam-mesos, take your time. Let us know you are satisfied, I'll do one last quick review & then merge.

elingg commented:
That is, take your time, but of course, the sooner the better! ;)

adam-mesos commented on conf/mesos-site.xml:
```diff
@@ -19,54 +19,84 @@
 
   <property>
     <name>mesos.hdfs.mesosdns.domain</name>
+    <description>Root domain name of Mesos DNS (usually 'Mesos')</description>
```
s/'Mesos'/'mesos'/ ?

adam-mesos replied:
Also, this can be ignored/commented-out if mesos.hdfs.mesosdns defaults to false.

adam-mesos commented on conf/mesos-site.xml:
```diff
@@ -19,54 +19,84 @@
   
   <property>
     <name>mesos.hdfs.secondary.data.dir</name>
+    <description>The secondary data directory in HDFS</description>
```
What is the secondary data directory though? Seems like this is somehow tied to dfs.domain.socket.path
(not a merge blocker, just a documentation ambiguity)

adam-mesos commented on conf/mesos-site.xml:
```diff
@@ -19,54 +19,84 @@
+
   <property>
     <name>mesos.hdfs.framework.mnt.path</name>
+    <description>Mount location (if above config is marked false)</description>
```
Rather than 'above config' please reference the actual config param mesos.hdfs.native-hadoop-binaries, in case these get reordered.

adam-mesos commented on conf/mesos-site.xml:
```diff
@@ -19,54 +19,84 @@
   <property>
     <name>mesos.hdfs.state.zk</name>
     <value>localhost:2181</value>
+    <description>Hostname and port of zookeeper node location</description>
```
Can this be plural, in the case of multiple zks?

abhay-agarwal replied:
That's in the quorum field. This is used by the framework to write into ZK, so you only need one. The quorum is what HDFS uses for its own zookeeper handling (and i'm testing right now whether it actually requires multiple unique DNS entries).

adam-mesos replied:
Sure, but another point of multiple zk nodes is that you have backup zks to contact if the first one is down. It's not just about a voting quorum.

abhay-agarwal replied:
looks like it can be a comma separated list. I'll add that point

adam-mesos commented on conf/mesos-site.xml:
```diff
@@ -19,54 +19,84 @@
   <property>
     <name>mesos.master.uri</name>
     <value>zk://localhost:2181/mesos</value>
+    <description>Zookeeper entry for mesos master location</description>
```
Can't this be a plural zk, or a manually entered master URI(s)?

adam-mesos commented on conf/mesos-site.xml:
```diff
@@ -19,54 +19,84 @@
 
   <property>
     <name>mesos.hdfs.framework.name</name>
+    <description>Your cluster name when accessing files (hdfs://YOUR_NAME)</description>
```
Also doubles as the framework name listed in the Mesos UI.

adam-mesos commented on conf/mesos-site.xml:
```diff
@@ -19,54 +19,84 @@
+  <!--
+  <property>
+    <name>mesos.hdfs.jvm.overhead</name>
+    <description>Multiplier added to resource reservation to account for JVM allocation</description>
```
s/added to/on/ since add != multiply

adam-mesos commented on example-conf/mesos-dns/mesos-site.xml:
```diff
@@ -0,0 +1,49 @@
+  <property>
+    <name>mesos.master.uri</name>
+    <value>zk://master.mesos:2181/mesos</value>
+    <description>The zookeeper state information to detect the leading mesos master.</description>
```
s/state information/entry/ or znode

adam-mesos commented on example-conf/mesos-dns/mesos-site.xml:
```diff
@@ -0,0 +1,49 @@
+  <property>
+    <name>mesos.hdfs.zkfc.ha.zookeeper.quorum</name>
+    <value>master0.mesos:2181,master1.mesos:2181,master2.mesos:2181</value>
+    <description>This is where you must individually list all your zookeeper znodes for Hadoop's HA feature. If you colocate mesos master and zookeeper znodes, you can use Mesos DNS for this.</description>
```
These are not znodes, but zk hosts/ports. A `znode` is a key in the `host:port/znode` namespace for a particular value in zk.

adam-mesos commented on conf/mesos-site.xml:
```diff
@@ -19,54 +19,84 @@
     <name>mesos.hdfs.state.zk</name>
-    <value>localhost:2181</value>
+    <description>Comma-separated hostname-port pairs of zookeeper node locations</description>
+    <value>master.mesos:2181</value>
```
This isn't the mesos-dns config file, so please go back to the localhost default.

abhay-agarwal commented:
Good catch with the incomplete commit..

adam-mesos commented on example-conf/mesosphere-dcos/mesos-site.xml:
```diff
@@ -0,0 +1,60 @@
+  <property>
+    <name>mesos.hdfs.preshared-hadoop-binaries</name>
+    <value>true</value>
+    <description>DCOS comes with a pre-shared HDFS configuration in a single-tenant environment</description>
```
s/pre-?shared/native/

abhay-agarwal replied:
*_*

adam-mesos replied:
I meant that "preshared" in the 'name' and "pre-shared" in the description should both be changed to "native" now that you've renamed the rest.

adam-mesos commented:
LGTM. Just fix the last "preshared" reference, and I declare it shippable. :+1: 
Not sure if @elingg wants to take another look too.

elingg commented:
Give me some time to give it a quick final look tomorrow ASAP. Then I will merge, assuming no unexpected major issues.  Thanks guys!

elingg commented on conf/mesos-site.xml:
```diff
@@ -19,55 +19,85 @@
+    <value>3</value>
+  </property>
+
+  <!-- Additional settings for fine-tuning -->
```
Let's list cpus and mem commented out here as well since these are often modified by our users

elingg replied:
Oh I misread this due to GH formatting, looking good :) I see they are all there.

elingg commented on README.md:
```diff
@@ -48,3 +51,10 @@ Resource Reservation Instructions (Optional)
+1. In Marathon (or your other long-running process monitor) stop the hdfs scheduler application
+2. Shutdown the hdfs framework in Mesos: `curl -d "frameworkId=YOUR_FRAMEWORK_ID" -X POST http://YOUR_MESOS_URL:5050/master/shutdown`
+3. Access your zookeeper instance: `/PATH/TO/zookeeper/bin/zkCli.sh`
+4. Remove hdfs-mesos framework state from zookeeper: `rmr /hdfs-mesos`
```
Could we add an instruction here to clear the data directory? If we are shutting down and clearing zookeeper, there will be no task reconciliation so there will be no recovery of data.

elingg replied:
Let's also add an explanation of this property, mesos.hdfs.native-hadoop-binaries in the readme.

abhay-agarwal replied:
It's already in the conf description tags

elingg replied:
True, but considering this is a pretty significant property for the user (whether or not the binaries with configs will get distributed on the slaves), I think we should highlight it as a quick line in the readme.

elingg commented:
Great work, @abhay-agarwal! :+1:  I left 3 small comments. Once you address those, feel free to merge.

abhay-agarwal commented:
Alright. Thanks for all the constructive comments. Merging now.
